## Lab 1 : Logistic Regression

### Objective:
To understand and apply the Machine Learning (ML) pipeline using Logistic Regression for a binary classification problem





#### Background
Artificial Intelligence(AI):It is the field of creating systems capable of performing tasks and operate similar to human intelligence

Machine Learning(ML): It is a subset of AI where a system learns pattern and trend from a data itself without explicit programming

Deep Learning(DL): It is a subset of ML that uses neural networks with multiple layers

Data Science: It involves collecting, cleaning,analyzing, and interpreting the data to extract meaningful information



#### Logistic Regression
Logistic Regression is a supervised machine learning algorithm used for classification problems. Unlike linear regression, which predicts continuous values it predicts the probability that an input belongs to a specific class.

It is used for binary classification where the output can be one of two possible categories such as Yes/No, True/False or 0/1.

It uses sigmoid function to convert inputs into a probability value between 0 and 1.


Types:

Binomial Logistic Regression: 
Used when the dependent variable has only two possible categories, such as Yes/No, Pass/Fail, or 0/1. It is the most common type and is used for binary classification tasks.


Multinomial Logistic Regression: 
Used when the dependent variable has three or more unordered categories. For example, classifying animals as cat, dog, or sheep. It extends logistic regression to handle multiple classes.


Ordinal Logistic Regression: 
Used when the dependent variable has three or more categories with a natural order, such as Low, Medium, and High. It considers the ranking of categories during prediction.


#### Dataset:
We will use Heart Disease dataset containing medical attributes related to cardiovascular health.

### Task 1: Logistic Regression with a Single Feature

To build a logistic regression model using only one input feature to predict whether a patient has heart disease.

Input Feature: Cholesterol

Target Variable: HeartDisease

#### 1. Data Retrieval and Collection

In [59]:
import pandas as pd

df = pd.read_csv("Heart_Disease_Prediction.csv")

print("Shape:", df.shape)

print("\nColumns:")
print(df.columns)

df.head()

Shape: (270, 14)

Columns:
Index(['Age', 'Sex', 'Chest pain type', 'BP', 'Cholesterol', 'FBS over 120',
       'EKG results', 'Max HR', 'Exercise angina', 'ST depression',
       'Slope of ST', 'Number of vessels fluro', 'Thallium', 'Heart Disease'],
      dtype='str')


,Age,Sex,Chest pain type,BP,Cholesterol,FBS over 120,EKG results,Max HR,Exercise angina,ST depression,Slope of ST,Number of vessels fluro,Thallium,Heart Disease
0,70,1,4,130,322,0,2,109,0,2.4,2,3,3,Presence
1,67,0,3,115,564,0,2,160,0,1.6,2,0,7,Absence
2,57,1,2,124,261,0,0,141,0,0.3,1,0,7,Presence
3,64,1,4,128,263,0,0,105,1,0.2,2,1,7,Absence
4,74,0,2,120,269,0,2,121,1,0.2,1,1,3,Absence


In [60]:
df.describe()

,Age,Sex,Chest pain type,BP,Cholesterol,FBS over 120,EKG results,Max HR,Exercise angina,ST depression,Slope of ST,Number of vessels fluro,Thallium
count,270.000000,270.000000,270.000000,270.000000,270.000000,270.000000,270.000000,270.000000,270.000000,270.00000,270.000000,270.000000,270.000000
mean,54.433333,0.677778,3.174074,131.344444,249.659259,0.148148,1.022222,149.677778,0.329630,1.05000,1.585185,0.670370,4.696296
std,9.109067,0.468195,0.950090,17.861608,51.686237,0.355906,0.997891,23.165717,0.470952,1.14521,0.614390,0.943896,1.940659
min,29.000000,0.000000,1.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.00000,1.000000,0.000000,3.000000
25%,48.000000,0.000000,3.000000,120.000000,213.000000,0.000000,0.000000,133.000000,0.000000,0.00000,1.000000,0.000000,3.000000
50%,55.000000,1.000000,3.000000,130.000000,245.000000,0.000000,2.000000,153.500000,0.000000,0.80000,2.000000,0.000000,3.000000
75%,61.000000,1.000000,4.000000,140.000000,280.000000,0.000000,2.000000,166.000000,1.000000,1.60000,2.000000,1.000000,7.000000
max,77.000000,1.000000,4.000000,200.000000,564.000000,1.000000,2.000000,202.000000,1.000000,6.20000,3.000000,3.000000,7.000000


This Heart Disease Dataset contains the attributes such as age, Sex, Chest pain type, BP, Cholesterol, FBS over 120, EKG results, Max HR, Exercise angina, ST depression, Slope of ST, Number of vessels fluro, Thallium, Heart Disease. 



#### 2. Data Cleaning

In [61]:
# Check Missing values
df.isnull().sum()


Age                        0
Sex                        0
Chest pain type            0
BP                         0
Cholesterol                0
FBS over 120               0
EKG results                0
Max HR                     0
Exercise angina            0
ST depression              0
Slope of ST                0
Number of vessels fluro    0
Thallium                   0
Heart Disease              0
dtype: int64

In [62]:
# Clean column names
df.columns = df.columns.str.strip()

# Clean target column safely
df['Heart Disease'] = df['Heart Disease'].astype(str).str.strip()

# Convert to binary
df['Heart Disease'] = df['Heart Disease'].map({
    'Presence': 1,
    'Absence': 0
})

# Drop invalid rows
df = df.dropna(subset=['Heart Disease'])

# Convert type
df['Heart Disease'] = df['Heart Disease'].astype(int)

print("Missing values in target:", df['Heart Disease'].isnull().sum())

print(df['Heart Disease'].value_counts())

Missing values in target: 0
Heart Disease
0    150
1    120
Name: count, dtype: int64


In [63]:
# Check invalid values
df = df[df['Cholesterol'] > 0]

In [64]:
#Check data types
df.dtypes

Age                          int64
Sex                          int64
Chest pain type              int64
BP                           int64
Cholesterol                  int64
FBS over 120                 int64
EKG results                  int64
Max HR                       int64
Exercise angina              int64
ST depression              float64
Slope of ST                  int64
Number of vessels fluro      int64
Thallium                     int64
Heart Disease                int64
dtype: object

#### 3. Feature Design

In [65]:
X = df[['Cholesterol']]
y = df['Heart Disease']

print("Feature Shape:", X.shape)
print("Target Shape:", y.shape)

Feature Shape: (270, 1)
Target Shape: (270,)


Cholesterol is a relevant predictor because it increases risk of heart blockage, high cholestrol causes plaque buildup in arteries

#### 4. Algorithm Selection

We choose Logistic Regression because it is specifically used for problems with binary classificaton and output probability between 0 and 1 so it works well for medical diagnosis problem 

#### 5. Loss Function Selection

We use Binary Cross Entropy (Log Loss):

Loss=−[ylog(p)+(1−y)log(1−p)]

It measures how well the predicted probabilities match the actual labels(0 or 1)

#### 6. Model Training

In [66]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# Split data
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Create model
model = LogisticRegression()

# Train model
model.fit(X_train, y_train)

print("Model Trained Successfully")

Model Trained Successfully


The model learns a relationship between cholesterol and heart disease by adjusting weights (coefficient and intercept) to minimize log loss.

#### 7. Model Evaluation

In [67]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.6111111111111112
Precision: 0.5
Recall: 0.23809523809523808
F1 Score: 0.3225806451612903

Confusion Matrix:
[[28  5]
 [16  5]]


### Task2: Logistic Regression with Multiple Features
To build a classification model using all available input features to predict heart disease.

Input Features: All variables except HeartDisease

Target Variable: HeartDisease


#### 3. Feature Design 

In [68]:
from sklearn.preprocessing import StandardScaler

# Features (ALL except target)
X = df.drop('Heart Disease', axis=1)
y = df['Heart Disease']

# Encode categorical-like binary columns if needed (already numeric in this dataset)
# So we directly scale features

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

print("Feature shape:", X.shape)
print("Target shape:", y.shape)

Feature shape: (270, 13)
Target shape: (270,)


Using multiple features improves prediction because heart disease depends on many physiological factors, not just cholesterol

#### 4. Algorithm Selection
Logistic Regression is used because:

It works for binary classification

It handles multiple features well

#### 5. Loss Function Selection
Binary Cross-Entropy (Log Loss) is used:

Loss=−[ylog(p)+(1−y)log(1−p)]

It improves predictions by penalizing wrong confident predictions.



#### 6. Model Training

In [69]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled,
    y,
    test_size=0.2,
    random_state=42
)

model2 = LogisticRegression(max_iter=1000)

model2.fit(X_train, y_train)

print("Model 2 Trained Successfully")

Model 2 Trained Successfully


#### 7. Model Evaluation

In [70]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

y_pred2 = model2.predict(X_test)

acc2 = accuracy_score(y_test, y_pred2)
prec2 = precision_score(y_test, y_pred2)
rec2 = recall_score(y_test, y_pred2)
f12 = f1_score(y_test, y_pred2)

print("Accuracy:", acc2)
print("Precision:", prec2)
print("Recall:", rec2)
print("F1 Score:", f12)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred2))

Accuracy: 0.9074074074074074
Precision: 0.9
Recall: 0.8571428571428571
F1 Score: 0.8780487804878049

Confusion Matrix:
[[31  2]
 [ 3 18]]


#### MODEL COMPARISON

Task 1 (Single Feature: Cholesterol) uses only one input variable to predict heart disease, making it simple but limited in capturing overall patient health conditions.

Task 2 (Multiple Features) uses all available medical attributes, allowing the model to learn richer patterns in the dataset.

Performance Comparison

Task 2 shows higher accuracy, precision, recall, and F1-score compared to Task 1.

Task 1 performs relatively lower because a single feature cannot fully represent the complexity of heart disease.

Why Task 2 Performs Better

Heart disease depends on multiple factors (age, BP, ECG, heart rate, etc.).

More features provide better information for learning patterns.

Trade-offs

Task 1: Simple and highly interpretable, but less accurate.

Task 2: More accurate and robust, but less interpretable and slightly more complex.


### Discussion

In this session, We studied and analyse the logistic regression and applied it. We studied how logistic regression behaves under two different feature settings. Using only Cholesterol (Task 1) gives a simple model but limits the ability to capture the full medical condition of a patient.
Using all features (Task 2) significantly improves prediction because heart disease depends on multiple health indicators, not just one.

### Conclusion
Logistic Regression was successfully applied to predict heart disease using both single-feature and multi-feature approaches. Task 1 demonstrated simplicity and interpretability, while Task 2 demonstrated higher accuracy and better overall performance.